backend/entityservice/tasks/encoding_uploading.py

```diff
@@ -1,66 +1,76 @@
-from entityservice.utils import iterable_to_stream, fmt_bytes, clks_uploaded_to_project
+from entityservice.utils import fmt_bytes, clks_uploaded_to_project
 
 
 @celery.task(base=TracedTask, ignore_result=True, args_as_tags=('project_id', 'dp_id'))
 def handle_raw_upload(project_id, dp_id, receipt_token, parent_span=None):
-    # User has uploaded base64 encodings as JSON
+    """
+    User has uploaded base64 encodings as JSON, this task needs to copy the data into
+    our internal binary format.
+    """
     log = logger.bind(pid=project_id, dp_id=dp_id)
     log.info("Handling user provided base64 encodings")
 
     with DBConn() as db:
         if not check_project_exists(db, project_id):
             log.info("Project deleted, stopping immediately")
             return
-        expected_count = get_number_of_hashes(db, dp_id)
+        # Get number of blocks + total number of encodings from database
+        expected_count, block_count = get_encoding_metadata(db, dp_id)
 
-    log.info(f"Expecting to handle {expected_count} encodings")
+    log.info(f"Expecting to handle {expected_count} encodings in {block_count} blocks")
     mc = connect_to_object_store()
-
-    #### GLUE CODE
     raw_file = Config.RAW_FILENAME_FMT.format(receipt_token)
     raw_data = mc.get_object(Config.MINIO_BUCKET, raw_file)
-    data = json.loads(raw_data.data.decode('utf-8'))
-    if 'clks' not in data:
-        raise ValueError('can only handle CLKs at the moment.')
-    binary_data = b'\n'.join(''.join(clk.split('\n')).encode() for clk in data['clks']) + b'\n'
-    buffer = io.BytesIO(binary_data)
-    #### END GLUE
-
-    # Set up streaming processing pipeline
-    buffered_stream = iterable_to_stream(buffer)
-    text_stream = io.TextIOWrapper(buffered_stream, newline='\n')
-
-    first_hash_bytes = deserialize_bytes(next(text_stream))
-    uploaded_encoding_size = len(first_hash_bytes)
-
-    def filter_generator():
-        i = 0
-        yield i, first_hash_bytes
-        for i, line in enumerate(text_stream, start=1):
-            hash_bytes = deserialize_bytes(line)
-            if len(hash_bytes) != uploaded_encoding_size:
-                raise ValueError("Encodings were not all the same size")
-            yield i, hash_bytes
-
-        log.info(f"Processed {i + 1} hashes")
+
+    with opentracing.tracer.start_span('upload-encodings-to-db', child_of=parent_span):
+        # stream encodings with block ids from uploaded file
+        # convert each encoding to our internal binary format
+        # output into database for each block (temp or direct to minio?)
+        pipeline = convert_encodings_from_base64_to_binary(stream_json_clksnblocks(raw_data))
+        with DBConn() as db:
+            for entity_id, encoding_data, blocks in pipeline:
+                # write this encoding to files or database
+                insert_encoding_into_blocks(db, dp_id, blocks, entity_id, encoding_data)
```

Comment:
It would be fair to say this should be improved. Doing insertions one at a time (in fact 2 SQL queries per encoding) is sub-optimal.